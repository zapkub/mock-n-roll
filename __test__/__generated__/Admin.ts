import { isEqual } from "lodash";

export class __mock__Admin {
    called: any[] = [];

    on(name: string, ...args) {
        const result = this.called.find(c => {
            const calls = [name, ...args]
            return isEqual(c[0], calls)
        })
        if (!result) {
            throw new Error(`call ${name} with ${JSON.stringify(args)} does not exists`)
        }
        return result[1]
    }

    mocks = {
        setRole: () => {
            return {
                toReturn: (returnArg: string) => { this.called.push([["setRole",], returnArg]) }
            }
        }
    };

    setRole(): string {
        return this.on("setRole")
    }

    role: string;
}
