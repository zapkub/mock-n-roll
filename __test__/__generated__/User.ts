import { isEqual } from "lodash";

export class User {
    mocks = {};
    called: any[] = [];

    on(name: string, ...args) {
        const result = this.called.find(c => {
            const calls = [name, ...args]
            return isEqual(c[0], calls)
        })
        if (!result) {
            throw new Error(`call ${name} with ${JSON.stringify(args)} is not exists`)
        }
        return result[1]
    }
}
