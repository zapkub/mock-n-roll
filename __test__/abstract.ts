import { AccessorDeclaration } from 'ts-morph'
import { bind } from 'jest-each';
export interface User {
  id: string
  name: string
}
export interface Admin extends User {
  role: string

  setRole(): string
}

export interface RoleInfo {
  name: string
}

export type CreateInput = Pick<User, 'name'>
export type CreateManyInput = Pick<User, 'name'>
export type CreateAdminInput = Pick<Admin, 'name' | 'role'>

export class Elastic {
  init() {

  }
  searchUser(): User {
    return {
      id: 'TEST',
      name: 'TEST'
    }
  }
}

export interface UserRepository {

  init()

  bootstrap(): void

  bind(input: Promise<Elastic>): void

  randomUser(): User
  createUser(input: CreateInput): Promise<User>
  createManyUser(...inputs: CreateManyInput[]): Promise<User[]>
  createManyAdmin(...inputs: CreateAdminInput[]): Promise<Admin[]>

  gen: (input: string) => User

  defaultRole: string
  generic: <T>(input: T) => T
  delimiter(d: AccessorDeclaration): AccessorDeclaration
  delime: (d: { role: RoleInfo }) => AccessorDeclaration
}
