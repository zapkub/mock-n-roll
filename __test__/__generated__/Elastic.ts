import { isEqual } from "lodash";
import { User } from "../abstract";

export class __mock__Elastic {
    called: any[] = [];

    on(name: string, ...args: any) {
        const result = this.called.find(c => {
            const calls = [name, ...args]
            return isEqual(c[0], calls)
        })
        if (!result) {
            throw new Error(`call ${name} with ${JSON.stringify(args)} does not exists`)
        }
        return result[1]
    }

    mocks = {
        init: () => {
            return {
                toReturn: () => { this.called.push([["init",]]) }
            }
        },
        searchUser: () => {
            return {
                toReturn: (returnArg: User) => { this.called.push([["searchUser",], returnArg]) }
            }
        }
    };

    init() {
        return this.on("init")
    }

    searchUser(): User {
        return this.on("searchUser")
    }
}
