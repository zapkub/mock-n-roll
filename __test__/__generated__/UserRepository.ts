import { isEqual } from "lodash";
import { User, CreateInput, CreateManyInput, CreateAdminInput, Admin } from "../abstract";

export class UserRepository {
    mocks = {
        randomUser: () => {
            return {
                toReturn: (returnArg: User) => {
                    this.called.push([["randomUser",], returnArg])
                }
            }        
},
        createUser: (input: CreateInput) => {
            return {
                toReturn: (returnArg: Promise<User>) => {
                    this.called.push([["createUser", input], returnArg])
                }
            }        
},
        createManyUser: (...inputs: CreateManyInput[]) => {
            return {
                toReturn: (returnArg: Promise<User[]>) => {
                    this.called.push([["createManyUser", ...inputs], returnArg])
                }
            }        
},
        createManyAdmin: (...inputs: CreateAdminInput[]) => {
            return {
                toReturn: (returnArg: Promise<Admin[]>) => {
                    this.called.push([["createManyAdmin", ...inputs], returnArg])
                }
            }        
}
    };
    called: any[] = [];

    on(name: string, ...args) {
        const result = this.called.find(c => {
            const calls = [name, ...args]
            return isEqual(c[0], calls)
        })
        if (!result) {
            throw new Error(`call ${name} with ${JSON.stringify(args)} does not exists`)
        }
        return result[1]
    }

    randomUser(): User {
        return this.on("randomUser")
    }

    createUser(input: CreateInput): Promise<User> {
        return this.on("createUser", input)
    }

    createManyUser(...inputs: CreateManyInput[]): Promise<User[]> {
        return this.on("createManyUser", ...inputs)
    }

    createManyAdmin(...inputs: CreateAdminInput[]): Promise<Admin[]> {
        return this.on("createManyAdmin", ...inputs)
    }
}
