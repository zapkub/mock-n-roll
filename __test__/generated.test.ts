import { __mock__UserRepository } from './__generated__/UserRepository'
import { UserRepository, Elastic } from './abstract'
import { __mock__Elastic } from './__generated__/Elastic'

describe('Generated mock from class should work', () => {
  it('Should work', () => {
    function x(e: Elastic) {
      expect(e.searchUser()).toBeDefined()
    }

    const defaultElasic = new Elastic()
    const mockElastic = new __mock__Elastic()
    mockElastic.mocks.searchUser().toReturn({
      id: 'unit',
      name: 'unit'
    })

    x(defaultElasic)
    x(mockElastic)
  })
})

describe('Generated result testing', () => {
  const u = new __mock__UserRepository()

  it('should able to mock func arg', () => {
    u.mocks
      .callback(
        () =>
          function() {
            return 'x'
          }
      )
      .toReturn('y')
    expect(
      u.callback(
        () =>
          function() {
            return 'x'
          }
      )
    ).toEqual('y')

    u.mocks
      .callback(function() {
        return 'x'
      })
      .toReturn('z')
    expect(
      u.callback(function() {
        return 'x'
      })
    ).toEqual('z')
  })

  it('Should inheritance of source interface', async () => {
    u.defaultRole = 'user'
    function x(userRepository: UserRepository) {
      expect(userRepository.defaultRole).toEqual('user')
    }
    x(u)
  })

  it('Should return promise of array type correctly', async () => {
    u.mocks.createManyAdmin({ name: 'TEST', role: 'role' }).toReturn(([]))
    const result = await u.createManyAdmin({ name: 'TEST', role: 'role' })
    expect(result).toEqual([])
  })
  it('should able to mock FunctionType property', async () => {
    u.mocks.gen('Test').toReturn({ name: 'ba', id: 'foo' })
    const result = u.gen('Test')
    expect(result).toEqual({ name: 'ba', id: 'foo' })
  })
  it('Should return promise of object type correctly', async () => {
    u.mocks.randomUser().toReturn({ id: 'TEST', name: 'TEST' })
    const result = u.randomUser()
    expect(result).toEqual({ id: 'TEST', name: 'TEST' })
  })
  it('should return multiple mock correctly', async () => {
    u.mocks.createUser({ name: 'create_TEST_2' }).toReturn({
      id: 'TEST_2',
      name: 'TEST_2'
    })
    u.mocks.createUser({ name: 'create_TEST' }).toReturn({
      id: 'TEST',
      name: 'TEST'
    })
    let result = await u.createUser({ name: 'create_TEST' })
    expect(result).toEqual({
      id: 'TEST',
      name: 'TEST'
    })

    result = await u.createUser({ name: 'create_TEST_2' })
    expect(result).toEqual({
      id: 'TEST_2',
      name: 'TEST_2'
    })
  })

  it('called should match snapshot', () => {
    expect(u.called).toMatchSnapshot()
  })
})
