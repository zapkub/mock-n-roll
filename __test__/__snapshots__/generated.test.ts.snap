// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated result testing called should match snapshot 1`] = `
Array [
  Array [
    Array [
      "randomUser",
      "{}",
    ],
    Array [],
    Object {
      "id": "TEST",
      "name": "TEST",
    },
  ],
  Array [
    Array [
      "createManyAdmin",
      "{\\"0\\":{\\"name\\":\\"TEST\\",\\"role\\":\\"role\\"}}",
    ],
    Array [
      Object {
        "name": "TEST",
        "role": "role",
      },
    ],
    Promise {},
  ],
  Array [
    Array [
      "callback",
      "{\\"0\\":\\"_$$ND_FUNC$$_() => function () {\\\\n            return 'x';\\\\n        }\\"}",
    ],
    Array [
      [Function],
    ],
    "y",
  ],
  Array [
    Array [
      "callback",
      "{\\"0\\":\\"_$$ND_FUNC$$_function () {\\\\n            return 'x';\\\\n        }\\"}",
    ],
    Array [
      [Function],
    ],
    "z",
  ],
  Array [
    Array [
      "gen",
      "{\\"0\\":\\"Test\\"}",
    ],
    Array [
      "Test",
    ],
    Object {
      "id": "foo",
      "name": "ba",
    },
  ],
  Array [
    Array [
      "createUser",
      "{\\"0\\":{\\"name\\":\\"create_TEST_2\\"}}",
    ],
    Array [
      Object {
        "name": "create_TEST_2",
      },
    ],
    Promise {},
  ],
  Array [
    Array [
      "createUser",
      "{\\"0\\":{\\"name\\":\\"create_TEST\\"}}",
    ],
    Array [
      Object {
        "name": "create_TEST",
      },
    ],
    Promise {},
  ],
]
`;
